src/algebra/algebra/basic.lean

```diff
@@ -1449,54 +1449,72 @@ restrict_scalars.is_scalar_tower R A V
 
 end type_synonym
 
+/-! TODO: The following lemmas no longer involve `algebra` at all, and could be moved elsewhere. -/
 section semimodule
 open semimodule
 
-variables (R A M N : Type*) [comm_semiring R] [semiring A] [algebra R A]
-variables [add_comm_monoid M] [semimodule R M] [semimodule A M] [is_scalar_tower R A M]
-variables [add_comm_monoid N] [semimodule R N] [semimodule A N] [is_scalar_tower R A N]
+variables (S R M N : Type*) [semiring S] [semiring R] [has_scalar S R]
```

Comment:
It is mildly irritating that there is a `has_scalar S R` instance here, where usually we assume `has_scalar R S` (or `algebra R S`), see e.g. `monoid_algebra.lean`, `linear_algebra/matrix.lean`, `algebra/tower.lean`, `finsupp.smul_comm_class`, ...

If we're going to move the definitions anyway, it will blow away the history anyway, so there is no need to preserve the previous variable names.

Reply:
My thinking here was to have the "main" submodule continue to be a `submodule R M`, rather than switching to `submodule R S` here.

I can use `R'` instead of `S` if that's less objectionable?

Reply:
> it will blow away the history anyway

My intent in modifying these in place was to not blow away history and change type-class paths in the same PR; just in case this breaks someone and they end up having to bisect.

Reply:
> My thinking here was to have the "main" submodule continue to be a `submodule R M`, rather than switching to `submodule S M` here.

I have to admit I don't see why this is important, but `has_scalar R' R` is fine to me as well.

Reply:
> My intent in modifying these in place was to not blow away history and change type-class paths in the same PR; just in case this breaks someone and they end up having to bisect.

When I find the PR that breaks my code in bisection, I look at the discussion on github. So as long as the change happens in multiple commits in the PR, I can figure out what happened. But I see how this can be annoying to do.

How about this: change the `TODO: move this` comment to something like `TODO: move this to a suitable file once `linear_map.ker` is split from `linear_algebra/basic.lean`, so that we can tell in the future if we can do the move yet.